Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs

```diff
@@ -0,0 +1,263 @@
+        public Task<GoogleCredential> GetDefaultCredentialAsync()
+        {
+            return cachedCredentialTask.Value;
+        }
+
+        /// <summary>Creates a new default credential.</summary>
+        private async Task<GoogleCredential> CreateDefaultCredentialAsync()
+        {
+            // 1. First try the environment variable.
+            string credentialPath = GetEnvironmentVariable(CredentialEnvironmentVariable);
+            if (!String.IsNullOrWhiteSpace(credentialPath))
+            {
+                try
+                {
+                    return CreateDefaultCredentialFromFile(credentialPath);
+                }
+                catch (Exception e)
+                {
+                    // Catching generic exception type because any corrupted file could manifest in different ways including 
+                    // but not limited to the System, System.IO or from the Newtonsoft.Json namespace.
+                    throw new InvalidOperationException(String.Format("Error reading credential file from location {0}: {1}"
+                                            + "\nPlease check the value of the Environment Variable {2}",
+                                            credentialPath,
+                                            e.Message,
+                                            CredentialEnvironmentVariable));
+                }
+            }
+            
+            // 2. Then try the well known file.
+            credentialPath = GetWellKnownCredentialFilePath();
+            if (!String.IsNullOrWhiteSpace(credentialPath))
+            {
+                try
+                {
+                    return CreateDefaultCredentialFromFile(credentialPath);
+                }
+                catch (FileNotFoundException)
+                {
+                    // File is not present, eat the exception and move on to the next check.
+                    Logger.Debug(String.Format("Well-known credential file {0} not found.", credentialPath));
+                }
+                catch (Exception e)
+                {
+                    throw new InvalidOperationException(String.Format("Error reading credential file from location {0}: {1}"
+                                        + "\nPlease rerun 'gcloud auth login' to regenerate credentials file.",
+                                        credentialPath,
+                                        e.Message));
+                }
+            }
+
+            // 3. Then try the compute engine.     
+            Logger.Debug("Checking whether the application is running on ComputeEngine.");
+            if (await ComputeCredential.IsRunningOnComputeEngine().ConfigureAwait(false))
+            {
+                Logger.Debug("ComputeEngine check passed. Using ComputeEngine Credentials.");
+                return GoogleCredential.FromCredential(new ComputeCredential());
+            }
+
+            // If everything we could try has failed, report an error.
```

Comment:
My English isn't great, so I'll leave it to you, but I would change it to:
// If everything we tried has failed, throw an exception.